diff --git a/dest/key_store.js b/dest/key_store.js
index 6f0d8c7c56943c57f8ed867e18771d0f6801d815..28df2fc8318c7dd4034398b6b6bd3fe8b56fca79 100644
--- a/dest/key_store.js
+++ b/dest/key_store.js
@@ -82,32 +82,34 @@ import { KEY_PREFIXES, computeAppSecretKey, deriveKeys, derivePublicKeyFromSecre
    * @param contractAddress - The contract address to silo the secret key in the key validation request with.
    * @returns The key validation request.
    */ async getKeyValidationRequest(pkMHash, contractAddress) {
-        const [keyPrefix, account] = await this.getKeyPrefixAndAccount(pkMHash);
-        // Now we find the master public key for the account
-        const pkMBuffer = await this.#keys.getAsync(`${account.toString()}-${keyPrefix}pk_m`);
-        if (!pkMBuffer) {
-            throw new Error(`Could not find ${keyPrefix}pk_m for account ${account.toString()} whose address was successfully obtained with ${keyPrefix}pk_m_hash ${pkMHash.toString()}.`);
-        }
-        const pkM = Point.fromBuffer(pkMBuffer);
-        const computedPkMHash = await pkM.hash();
-        if (!computedPkMHash.equals(pkMHash)) {
-            throw new Error(`Could not find ${keyPrefix}pkM for ${keyPrefix}pk_m_hash ${pkMHash.toString()}.`);
-        }
-        // Now we find the secret key for the public key
-        const skStorageSuffix = secretKeyStorageSuffix(keyPrefix);
-        const skMBuffer = await this.#keys.getAsync(`${account.toString()}-${skStorageSuffix}`);
-        if (!skMBuffer) {
-            throw new Error(`Could not find ${skStorageSuffix} for account ${account.toString()} whose address was successfully obtained with ${keyPrefix}pk_m_hash ${pkMHash.toString()}.`);
-        }
-        const skM = GrumpkinScalar.fromBuffer(skMBuffer);
-        // We sanity check that it's possible to derive the public key from the secret key
-        const derivedPkM = await derivePublicKeyFromSecretKey(skM);
-        if (!derivedPkM.equals(pkM)) {
-            throw new Error(`Could not derive ${keyPrefix}pkM from ${keyPrefix}skM.`);
-        }
-        // At last we silo the secret key and return the key validation request
-        const skApp = await computeAppSecretKey(skM, contractAddress, keyPrefix);
-        return new KeyValidationRequest(pkM, skApp);
+        return this.#db.transactionAsync(async () => {
+            const [keyPrefix, account] = await this.getKeyPrefixAndAccount(pkMHash);
+            // Now we find the master public key for the account
+            const pkMBuffer = await this.#keys.getAsync(`${account.toString()}-${keyPrefix}pk_m`);
+            if (!pkMBuffer) {
+                throw new Error(`Could not find ${keyPrefix}pk_m for account ${account.toString()} whose address was successfully obtained with ${keyPrefix}pk_m_hash ${pkMHash.toString()}.`);
+            }
+            const pkM = Point.fromBuffer(pkMBuffer);
+            const computedPkMHash = await pkM.hash();
+            if (!computedPkMHash.equals(pkMHash)) {
+                throw new Error(`Could not find ${keyPrefix}pkM for ${keyPrefix}pk_m_hash ${pkMHash.toString()}.`);
+            }
+            // Now we find the secret key for the public key
+            const skStorageSuffix = secretKeyStorageSuffix(keyPrefix);
+            const skMBuffer = await this.#keys.getAsync(`${account.toString()}-${skStorageSuffix}`);
+            if (!skMBuffer) {
+                throw new Error(`Could not find ${skStorageSuffix} for account ${account.toString()} whose address was successfully obtained with ${keyPrefix}pk_m_hash ${pkMHash.toString()}.`);
+            }
+            const skM = GrumpkinScalar.fromBuffer(skMBuffer);
+            // We sanity check that it's possible to derive the public key from the secret key
+            const derivedPkM = await derivePublicKeyFromSecretKey(skM);
+            if (!derivedPkM.equals(pkM)) {
+                throw new Error(`Could not derive ${keyPrefix}pkM from ${keyPrefix}skM.`);
+            }
+            // At last we silo the secret key and return the key validation request
+            const skApp = await computeAppSecretKey(skM, contractAddress, keyPrefix);
+            return new KeyValidationRequest(pkM, skApp);
+        });
     }
     /**
    * Gets the master nullifier public key for a given account.
@@ -194,18 +196,20 @@ import { KEY_PREFIXES, computeAppSecretKey, deriveKeys, derivePublicKeyFromSecre
    * @returns A Promise that resolves to sk_m.
    * @dev Used when feeding the sk_m to the kernel circuit for keys verification.
    */ async getMasterSecretKey(pkM) {
-        const [keyPrefix, account] = await this.getKeyPrefixAndAccount(pkM);
-        const skStorageSuffix = secretKeyStorageSuffix(keyPrefix);
-        const secretKeyBuffer = await this.#keys.getAsync(`${account.toString()}-${skStorageSuffix}`);
-        if (!secretKeyBuffer) {
-            throw new Error(`Could not find ${skStorageSuffix} for ${keyPrefix}pk_m ${pkM.toString()}. This should not happen.`);
-        }
-        const skM = GrumpkinScalar.fromBuffer(secretKeyBuffer);
-        const derivedpkM = await derivePublicKeyFromSecretKey(skM);
-        if (!derivedpkM.equals(pkM)) {
-            throw new Error(`Could not find ${skStorageSuffix} for ${keyPrefix}pkM ${pkM.toString()} in secret keys buffer.`);
-        }
-        return Promise.resolve(skM);
+        return this.#db.transactionAsync(async () => {
+            const [keyPrefix, account] = await this.getKeyPrefixAndAccount(pkM);
+            const skStorageSuffix = secretKeyStorageSuffix(keyPrefix);
+            const secretKeyBuffer = await this.#keys.getAsync(`${account.toString()}-${skStorageSuffix}`);
+            if (!secretKeyBuffer) {
+                throw new Error(`Could not find ${skStorageSuffix} for ${keyPrefix}pk_m ${pkM.toString()}. This should not happen.`);
+            }
+            const skM = GrumpkinScalar.fromBuffer(secretKeyBuffer);
+            const derivedpkM = await derivePublicKeyFromSecretKey(skM);
+            if (!derivedpkM.equals(pkM)) {
+                throw new Error(`Could not find ${skStorageSuffix} for ${keyPrefix}pkM ${pkM.toString()} in secret keys buffer.`);
+            }
+            return skM;
+        });
     }
     /**
    * Checks whether a given account has a key matching the provided master public key hash.
@@ -213,14 +217,16 @@ import { KEY_PREFIXES, computeAppSecretKey, deriveKeys, derivePublicKeyFromSecre
    * @param pkMHash - The master public key hash to look for.
    * @returns True if the account has a key with the given hash.
    */ async accountHasKey(account, pkMHash) {
-        const pkMHashBuffer = serializeToBuffer(pkMHash);
-        for (const prefix of KEY_PREFIXES){
-            const stored = await this.#keys.getAsync(`${account.toString()}-${prefix}pk_m_hash`);
-            if (stored && Buffer.from(stored).equals(pkMHashBuffer)) {
-                return true;
+        return this.#db.transactionAsync(async () => {
+            const pkMHashBuffer = serializeToBuffer(pkMHash);
+            for (const prefix of KEY_PREFIXES){
+                const stored = await this.#keys.getAsync(`${account.toString()}-${prefix}pk_m_hash`);
+                if (stored && Buffer.from(stored).equals(pkMHashBuffer)) {
+                    return true;
+                }
             }
-        }
-        return false;
+            return false;
+        });
     }
     /**
    * Gets the key prefix and account address for a given value.
